Implemented working Name tag

IvanCraft623 opened the pull request:
## Introduction
<!-- Explain existing problems or why this pull request is necessary -->
Implement the nametags and their function (currently only obtainable through commands because there are no anvils).

## Follow-up
- https://github.com/pmmp/Language/pull/134

## Tests

https://youtu.be/DCpjGbXHEh0
(The interactive tag for fishing rod was implemented overriding the item on a plugin for testing the api)

IvanCraft623 committed:
Implement Name Tag

IvanCraft623 committed:
Pop item

IvanCraft623 committed:
Oops

xerenahmed commented:
I think an event must be written for this feature.

IvanCraft623 commented:
> I think an event must be written for this feature.

There is PlayerEntityInteractEvent

dktapps commented on src/entity/Living.php:
```diff
@@ -352,6 +353,19 @@ protected function onHitGround() : ?float{
 
+	public function onInteract(Player $player, Vector3 $clickPos) : bool{
+		$item = $player->getInventory()->getItemInHand();
+		if($item->getTypeId() === ItemTypeIds::NAME_TAG && !$this instanceof Human && $item->hasCustomName()){
```
`$this instanceof` defeats the purpose of OOP. Also, I'm not sure this logic belongs in `Living`.

IvanCraft623 replied:
> I'm not sure this logic belongs in `Living`

According to my tests and the wiki, all Living entities with the exception of EnderDragon and Human can be named with the nametag.

IvanCraft623 replied:
> `$this instanceof` defeats the purpose of OOP

Solved.

IvanCraft623 committed:
Move check to Human class

IvanCraft623 committed:
Merge branch 'next-major' into name-tag

IvanCraft623 committed:
Merge branch 'next-major' into name-tag

IvanCraft623 committed:
Merge branch 'next-major' into name-tag

IvanCraft623 committed:
Merge branch 'next-major' into name-tag

IvanCraft623 committed:
Merge branch 'next-major' into name-tag

IvanCraft623 committed:
@IvanCraft623 Changes for new ItemSerializerDeserializerRegistrar

xRodzMC commented:
Looks good, well done

dktapps commented:
The PR looks functionally OK.

The thing that's giving me pause is that the mode of operation here is inconsistent with every other method of item interaction. For example, weapon damage when attacking an entity is not handled by the entity itself; it's processed in `Item->onAttackEntity()`. We also already have API methods like `Item->onInteractBlock()`.

IvanCraft623 commented:
> The PR looks functionally OK.
> 
> The thing that's giving me pause is that the mode of operation here is inconsistent with every other method of item interaction. For example, weapon damage when attacking an entity is not handled by the entity itself; it's processed in `Item->onAttackEntity()`. We also already have API methods like `Item->onInteractBlock()`.

I agree with adding a method like that, but for this specific case I am more inclined towards the idea of adding an interface for renameable entities because they have a property: `EntityMetadataProperties::INTERACTIVE_TAG`.

IvanCraft623 committed:
Implement interactive tag

IvanCraft623 committed:
Oops

IvanCraft623 committed:
Add more documentation

IvanCraft623 commented:
The following changes have been made:
 - Logic for renaming has been moved to `NameableTrait`, renameable entities should use it and implement `Nameable`
 - Implemented API for Interactive Tags (interaction button displayed above the hotbar).
 - Interactive tag is now displayed when looking at the renameable entity.

The above changes require [this](https://github.com/pmmp/Language/pull/134) to be merged

IvanCraft623 committed:
Vanilla checks entity before item

IvanCraft623 committed:
Merge branch 'next-major' into name-tag

IvanCraft623 committed:
Merge branch 'next-major' into name-tag

IvanCraft623 committed:
Use KnownTranslationKeys constant

IvanCraft623 committed:
Merge branch 'next-major' into name-tag

IvanCraft623 committed:
Use new Item method for entity interaction

IvanCraft623 commented:
This PR has caught up with the latest changes and now uses the new `Item::onInteractEntity()` method.

dktapps commented on src/entity/NameableTrait.php:
```diff
@@ -0,0 +1,45 @@
+			return KnownTranslationKeys::ACTION_INTERACT_NAME;
+		}
+		return parent::getInteractiveTag($player, $item);
+	}
```
This seems redundant. It can be implemented in Living directly.

dktapps commented on src/entity/Nameable.php:
```diff
@@ -0,0 +1,39 @@
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+
+interface Nameable{
```
This interface seems redundant.

dktapps commented on src/player/Player.php:
```diff
@@ -607,6 +611,36 @@ public function setDisplayName(string $name) : void{
+		}else{
+			$this->lookingId = $looking->getId();
+		}
+	}
```
This entire block looks redundant.

dktapps commented on src/player/Player.php:
```diff
@@ -1852,6 +1887,23 @@ public function interactEntity(Entity $entity, Vector3 $clickPos) : bool{
+			return true;
+		}
+		return false;
+	}
```
This also looks redundant.

IvanCraft623 replied:
What is redundant?

dktapps reviewed (dismissed):
The interactive tags stuff looks like a really overengineered mess and can be greatly simplified.

IvanCraft623 commented:
I do not understand what is redundant
`Nameable` and `NameableTrait` exists because not all `Living` entities are nameable, some examples are:
- `Human`
- `EnderDragon`
and there are also non-`Living` entities that are nameable:
- `ArmorStand`

IvanCraft623 committed:
Merge branch 'next-major' into name-tag

jasonw4331 commented:
I believe the look detection is what is being considered as redundant here. The nametag can be implemented from when it gets used and shouldn't depend on whether or not an entity is being looked at.

dktapps commented:
The look detection isn't redundant in itself, I just think that the implementation of it is vastly overcomplicated.

IvanCraft623 committed:
Merge branch 'minor-next' into name-tag

dktapps committed:
Merge branch 'minor-next' into name-tag

dktapps committed:
Strip out a bunch of unrelated junk

dktapps committed:
Added Entity->canBeRenamed()
since every entity can have a nametag anyway, the only real distinction to be made is which ones can be named by a nametag.
this way is also more dynamic since it can be overridden to create special conditions (e.g. perhaps certain subtypes of entities that can't be renamed)

dktapps committed:
Remove unrelated changes
